crates/sui-framework/sources/hex.move

```diff
@@ -23,6 +30,27 @@ module sui::hex {
         r
     }
 
+    /// Decode hex into `bytes`
+    public fun decode(hex: vector<u8>): vector<u8> {
+        let (i, r, l) = (0, vector[], vector::length(&hex));
+        while (i < l) {
+            let decimal = (base_16_index(vector::borrow(&hex, i)) * 16) + 
+                          base_16_index(vector::borrow(&hex, i + 1));
+            vector::push_back(&mut r, decimal);
+            i = i + 2;
+        };
+        r
+    }
+    
+    fun base_16_index(base_16_value: &u8): u8 {
+        let (found, base16) = vector::index_of(&BASE16, base_16_value);
+        assert!(found, ENotValidHexCharacter);
+        if (base16 >= 16) {
+            base16 = base16 - 6;
+        };
+        (base16 as u8)
+    }
```

Comment:
Rather than using a vector and `index_of`, can you use a series of range checks?

```suggestion
    fun decode_byte(hex: u8): u8 {
        if (/* 0 .. 9 */ 48 <= hex && hex < 58) {
            hex - 48
        } else if (/* A .. F */ 65 <= hex && hex < 71) {
            10 + hex - 65
        } else if (/* a .. f */ 97 <= hex && hex < 103) {
            10 + hex - 97
        } else {
            abort!(ENotValidHexCharacter);
        }
    }
```